### DISCUSSION TITLE
Change Object Parent / Key without Flushing Saved but Unpublished Changes

### DISCUSSION THREAD

**Discussion by cancan101** [pimcore/pimcore#16962] [Q&A / Support] [answered]
If I have "saved" but not "published" (ie in DB) changes to an Object, is there anyway to change an Object's Parent (or Key) without also flushing (i.e. saving) those other changes to the DB?

We would like to change an Object's Parent as it moves through a workflow (so that our users can easily find all of the Objects at a given state). However we are not publishing (to the DB) until the final state of the workflow. Unfortunately any change to the Parent field requires a "save", which then flushes all the other saved but unpublished changes to the DB. 

Is there any way around this, where I can selectively update in the DB only certain fields (ie the Parent)?

**Comment by markus-moser** [comment]
This is a quite complex question. There is definitely no way to selectively update single fields only. But when you load the current object from the database you will get the currently stored database state/entry. When you then only change the parent and save the object with disabled object versioning (Version::disable()) you will only change the parent in the database and not infect or create any versions. In the  unpublished version you would then need to ensure that it does not get moved back to the original folder as soon as it get's published.

When you want to do this some kind of event listener within the workflow you could do something like: `$object = AbstractObject->getById($object->getId(), true)` to force getting the database state of the object.

**Reply by cancan101** [reply]
If I use `Version::disable()`, it prevents a new version from being created, but it does cause some issue with  the "published" indicator on versions:
![image](https://github.com/pimcore/pimcore/assets/51059/de254054-303b-4f3f-8573-1f50914d3b1d)

I assume this has to do with the mod date, etc being changed? i.e. in that screenshot, the published indicator got cleared when `save` was called even with `Version::disable()` (no new version line was added though).

**Reply by markus-moser** [reply]
Yes this is because of the modification date. Not sure, but you could try to do a `$object->setModificationDate($object->getModificationDate())` to try to avoid changing of the modification date.

**Reply by cancan101** [reply]
> When you want to do this some kind of event listener within the workflow you could do something like: `$object = AbstractObject->getById($object->getId(), true)` to force getting the database state of the object.

Will calling `save` on that object cause Pimcore to purge the pending "unpublished" versions (from the filesystem)? Or will they still exist as unpublished changes sitting on top of the published Object?

**Reply by cancan101** [reply]
> Yes this is because of the modification date. Not sure, but you could try to do a `$object->setModificationDate($object->getModificationDate())` to try to avoid changing of the modification date.

Are there any other fields that I would want to prevent from changing (or overwrite back to the current value) if all I care about is changing the parent and don't want "anything else" to change?

**Reply by markus-moser** [reply]
Just calling `save()` should not load the version from the file system and persist it to the DB, the file system version should still exist and be ontouched.

**Reply by cancan101** [reply]
> Just calling `save()` should not load the version from the file system and persist it to the DB, the file system version should still exist and be ontouched.

What is the process in PImcore that causes Pimcore to cleanup the file system record? (ie to mark it as "applied")

**Reply by cancan101** [reply]
So far I still can't get the publish logo to show:

This is the logic used to determine if the publish icon should be shown:
```js
renderer: function (d, metaData, cellValues) {
    var d = Ext.Date.format(cellValues.get('date'), "timestamp");
    var versionCount = cellValues.get('versionCount');
    var index = cellValues.get('index');
    if (index === 0 && d == this.object.data.general.versionDate && versionCount == this.object.data.general.versionCount) {
        if(this.object.data.general.o_published) {
            metaData.tdCls = "pimcore_icon_publish";
        } else {
```
The mismatch is here: `d == this.object.data.general.versionDate`, so what causes that to be set?
I am currently using this:
```php
$object = DataObject\AbstractObject::getById($object->getId(), true);

$parentId = $parentObject->getId();
$object->setParentId($parentId);
$object->setKey($key);
$object->setModificationDate($object->getModificationDate());
$object->setUserModification($object->getUserModification());

\Pimcore\Model\Version::disable();
$object->save();
\Pimcore\Model\Version::enable();
```

**Reply by markus-moser** [reply, marked as answer]
This version date is based on the modification date. The reason why it does not work is that the modification date only is set as dirty when it actually changes: 

https://github.com/pimcore/pimcore/blob/11.x/models/Element/AbstractElement.php#L162

All of this seems to be really some kind of hack, but if you would like to achieve that the published checkbox is still there, you would need something like this:

```php
$object = DataObject\AbstractObject::getById($object->getId(), true);

$parentId = $parentObject->getId();
$object->setParentId($parentId);
$object->setKey($key);
$modificationDate = $object->getModificationDate();
$object->setModificationDate(0);
$object->setModificationDate($modificationDate);
$object->setUserModification($object->getUserModification());

\Pimcore\Model\Version::disable();
$object->save();
\Pimcore\Model\Version::enable();
```

**Reply by cancan101** [reply]
That worked! Is there anything I should be concerned with as far as the object cache?
https://github.com/pimcore/pimcore/blob/d44ada06332654086818bd7d6c64c8274af5705c/models/DataObject/AbstractObject.php#L738-L739

**Reply by markus-moser** [reply]
No the cache is no problem at all.

**Reply by cancan101** [reply]
For my understanding, what does the concept of a field being marked `dirty` mean?

**Reply by markus-moser** [reply]
`dirty` means that the field value got changed

**Reply by cancan101** [reply]
I understand that, but what is the is practical implication of a dirty field? i.e. why does Pimcore track dirty fields and what does it do with this information.

**Reply by markus-moser** [reply]
It for example avoids updating fields which are not changed in the database to reduce the amount of needed SQL statements and speed things up.

**Comment by markus-moser** [comment]
> What is the process in PImcore that causes Pimcore to cleanup the file system record? (ie to mark it as "applied")

The data object (so the database state) and it's version are two relatively independent things. The version is a php serialized version of the object on a given timestamp in the filesystem. So it contains all attributes 1:1 as they were when you saved the object and the version was created. So when you are working with the normal PHP API you will not get the latest version but the data object (database representation). The whole topic of "replacing" the data base state with information from the filesystem version is done in the Pimcore backend (see e.g. here https://github.com/pimcore/admin-ui-classic-bundle/blob/d191bcd0184f77e3ac9d25270ee02a4fcc2496aa/src/Controller/Admin/DataObject/DataObjectController.php#L302). So when you open the object you will see the latest version instead of the database content. When you then publish the object the version information will be persistet in the database.
